In [7]:
from pathlib import Path
import pandas as pd

## Loading the data


In [ ]:
data_path = Path(r"QuestionDataV2_06122025_classified.csv").resolve()
df = pd.read_csv(data_path)
print(df.columns)

Index(['Unnamed: 0', 'Question Title', 'info.json', 'question.html',
       'server.js', 'solution.html', 'server.py', 'properties.js', 'question',
       'uuid', 'title', 'stem', 'topic', 'tags', 'prereqs', 'isAdaptive',
       'createdBy', 'qType', 'nSteps', 'updatedBy', 'difficulty', 'codelang',
       'resources', 'stepType', 'dificulty', 'relevant_courses', 'topics'],
      dtype='object')


## Inspecting


In [10]:
df.head()

,Unnamed: 0,Question Title,info.json,question.html,server.js,solution.html,server.py,properties.js,question,uuid,...,qType,nSteps,updatedBy,difficulty,codelang,resources,stepType,dificulty,relevant_courses,topics
0,0,3dMoment1,"{\r\n ""uuid"": ""8abcd91d6-64d1-4e89-b72d-ad2...",<pl-question-panel>\r\n <pl-figure file-nam...,const math = require('mathjs');\r\n// const ma...,<pl-solution-panel>\r\n <pl-figure file-nam...,NaN,NaN,determine the moment about point o in ($n \cdo...,8abcd91d6-64d1-4e89-b72d-ad2133f25b2f,...,Num,1,NaN,1.0,Javascript,NaN,NaN,NaN,"ME 010, ME 103","Static Equilibrium, Distributed Forces"
1,1,3dMoment2,"{\r\n ""uuid"": ""8abcd91d6-64d1-4e89-b72d-ad2...",<pl-question-panel>\r\n <pl-figure file-nam...,const math = require('mathjs');\r\n// const ma...,<pl-solution-panel>\r\n <pl-figure file-nam...,NaN,NaN,"Refer to the diagram labeled ""3dMoment2.png"". ...",8abcd91d6-64d1-4e89-b72d-ad2133f25b2f,...,Num,1,NaN,1.0,Javascript,NaN,NaN,NaN,"ME 103, ME 110",3D Rigid Body Mechanics
2,2,3dMoment3,"{\r\n ""uuid"": ""8abcd91d6-64d1-4e89-b72d-ad2...",<pl-question-panel>\r\n <pl-figure file-nam...,const math = require('mathjs');\r\n// const ma...,<pl-solution-panel>\r\n <pl-figure file-nam...,NaN,NaN,determine the resultant moment produced by the...,8abcd91d6-64d1-4e89-b72d-ad2133f25b2f,...,Num,1,NaN,1.0,Javascript,NaN,NaN,NaN,NaN,"Static Equilibrium, Distributed Forces"
3,3,3dMoment4,"{\r\n ""uuid"": ""8abcd91d6-64d1-4e89-b72d-ad2...",<pl-question-panel>\r\n <pl-figure file-nam...,const math = require('mathjs');\r\n// const ma...,<pl-solution-panel>\r\n <pl-figure file-nam...,NaN,NaN,determine the resultant moments produced by th...,8abcd91d6-64d1-4e89-b72d-ad2133f25b2f,...,Num,1,NaN,1.0,Javascript,NaN,NaN,NaN,"ME 010, ME 103","Statics and Mechanics, Static Equilibrium, Vec..."
4,4,Equilibrium1,"{\r\n ""uuid"": ""8abcd91d6-64d1-4e89-b72d-ad2...","<pl-question-panel>\r\n<pl-figure file-name=""3...",const math = require('mathjs');\r\n// const ma...,<pl-solution-panel>\r\n <pl-figure file-nam...,NaN,NaN,compute the missing force vector that will all...,8abcd91d6-64d1-4e89-b72d-ad2133f25b2f,...,Num,1,NaN,1.0,Javascript,NaN,NaN,NaN,ME 010,"Static Equilibrium, Vector Analysis"


In [ ]:
columns_to_keep = [
    "Question Title",
    "question.html",
    "server.js",
    "solution.html",
    "server.py",
    "properties.js",
    "info.json",
    "isAdaptive",
]
df = df[columns_to_keep]
df.head()

,Question Title,question.html,server.js,solution.html,server.py,properties.js,info.json,isAdaptive
0,3dMoment1,<pl-question-panel>\r\n <pl-figure file-nam...,const math = require('mathjs');\r\n// const ma...,<pl-solution-panel>\r\n <pl-figure file-nam...,NaN,NaN,"{\r\n ""uuid"": ""8abcd91d6-64d1-4e89-b72d-ad2...",True
1,3dMoment2,<pl-question-panel>\r\n <pl-figure file-nam...,const math = require('mathjs');\r\n// const ma...,<pl-solution-panel>\r\n <pl-figure file-nam...,NaN,NaN,"{\r\n ""uuid"": ""8abcd91d6-64d1-4e89-b72d-ad2...",True
2,3dMoment3,<pl-question-panel>\r\n <pl-figure file-nam...,const math = require('mathjs');\r\n// const ma...,<pl-solution-panel>\r\n <pl-figure file-nam...,NaN,NaN,"{\r\n ""uuid"": ""8abcd91d6-64d1-4e89-b72d-ad2...",True
3,3dMoment4,<pl-question-panel>\r\n <pl-figure file-nam...,const math = require('mathjs');\r\n// const ma...,<pl-solution-panel>\r\n <pl-figure file-nam...,NaN,NaN,"{\r\n ""uuid"": ""8abcd91d6-64d1-4e89-b72d-ad2...",True
4,Equilibrium1,"<pl-question-panel>\r\n<pl-figure file-name=""3...",const math = require('mathjs');\r\n// const ma...,<pl-solution-panel>\r\n <pl-figure file-nam...,NaN,NaN,"{\r\n ""uuid"": ""8abcd91d6-64d1-4e89-b72d-ad2...",True


### IsAdaptive Check


In [ ]:
row_num, col_num = df.shape
adaptive_questions_count = df["isAdaptive"].sum()
print(
    f"""The number of adaptive questions is {adaptive_questions_count} \n while non adaptive questions are {row_num-adaptive_questions_count}
"""
)
py_filtered = df[(df["isAdaptive"]) & (df["server.py"].notna())]
js_filtered = df[(df["isAdaptive"]) & (df["server.js"].notna())]

The number of adaptive questions is 158 
 while non adaptive questions are 105



In [23]:
import re
import json
pattern = r"</pl-[a-z\-]+>"
question_examples = {}
for text in df["question.html"]:
    result = re.findall(pattern, text)
    for r in result:
        if not question_examples.get(r, None):
            question_examples[r] = text

save_path = Path(r"example_questions.json").resolve()
print(save_path)
save_path.write_text(json.dumps(question_examples))
print(question_examples)

C:\Github\GestatlQuestions\backend\src\data\example_questions.json
{'</pl-figure>': '<pl-question-panel>\r\n    <pl-figure file-name="3dMoment1.png"></pl-figure>\r\n   <p>Determine the moment about point O in (${{params.unitsForce}} {{params.unitsDist}}$) caused by the force $ \\mathbf{F} = {{params.force}}$ $ {{params.unitsForce}}$ acting on the beam.</p>\r\n   <p>Let the lengths $\\mathbf{L}_1 = {{params.l1}}$ $ {{params.unitsDist}} $, $\\mathbf{L}_2 = {{params.l2}}$ $ {{params.unitsDist}} $, $\\mathbf{L}_3 = {{params.l3}}$ $ {{params.unitsDist}} $</p>\r\n   <p>and $\\mathbf{V}_1 = {{params.v1}}$ $ {{params.unitsDist}} $, $\\mathbf{V}_2 = {{params.v2}}$ $ {{params.unitsDist}} $, $\\mathbf{V}_3 = {{params.v3}}$ $ {{params.unitsDist}} $</p>\r\n</pl-question-panel>\r\n\r\n<pl-number-input answers-name="momenti" comparison="sigfig" digits="3" label="Moment about point O $\\hat{i}$ ="></pl-number-input>\r\n<pl-number-input answers-name="momentj" comparison="sigfig" digits="3" label="Momen

## Notes

- info.json: Originally held question metadata, now empty. Not essential for generation but should be repopulated.
- question.html: HTML template that renders the student-facing question (figures, inputs, choices, etc.).
- server files: Backend logic written in either JavaScript or Python.
  -Total questions: 320 (used for semantic search); not guaranteed to reflect the current total.
- Adaptive questions: 158 — dynamically generated (e.g. numerical computations).
- Non-adaptive questions: 162 — static (e.g. multiple choice).
- Code files: 7 Python files, 139 JavaScript files related to questions.
